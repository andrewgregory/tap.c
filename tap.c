/*
 * Copyright 2014 Andrew Gregory andrew.gregory.8@gmail.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include <string.h>
#include <stdio.h>
#include <math.h>
#include <stdarg.h>

#ifndef __TAP_C
#define __TAP_C

int _tap_tests_run = 0;
int _tap_tests_failed = 0;

const char *_tap_todo = NULL;

void tap_plan(int test_count)
{
    printf("1..%d\n", test_count);
}

void tap_skip_all(const char *reason, ...)
{
    printf("1..0");
    if(reason) {
        va_list args;
        va_start(args, reason);
        fputc(' ', stdout);
        vprintf(reason, args);
        va_end(args);
    }
    fputc('\n', stdout);
}

void tap_done_testing(void)
{
    tap_plan(_tap_tests_run);
}

/*

int tap_get_tests_run(void)
{
    return _tap_tests_run;
}

int tap_get_tests_failed(void)
{
    return _tap_tests_failed;
}

*/

void tap_todo(const char *reason)
{
    _tap_todo = reason;
}

void tap_skip(int count, const char *reason, ...)
{
    while(count--) {
        printf("ok %d # SKIP", ++_tap_tests_run);
        if(reason) {
            va_list args;
            fputs(" - ", stdout);
            va_start(args, reason);
            vprintf(reason, args);
            va_end(args);
        }
        fputc('\n', stdout);
    }
}

void tap_bail(const char *reason, ...)
{
    va_list args;
    fputs("Bail out! ", stdout);
    va_start(args, reason);
    vprintf(reason, args);
    va_end(args);
    fputc('\n', stdout);
}

void tap_diag(const char *message, ...)
{
    va_list args;
    fputs("    # ", stdout);
    va_start(args, message);
    vprintf(message, args);
    va_end(args);
    fputc('\n', stdout);
}

void _tap_vok(int success, const char *name, va_list args)
{
    const char *result;
    if(success) {
        result = "ok";
        if(_tap_todo) ++_tap_tests_failed;
    } else {
        result = "not ok";
        if(!_tap_todo) ++_tap_tests_failed;
    }

    printf("%s %d", result, ++_tap_tests_run);

    if(name) {
        fputs(" - ", stdout);
        vprintf(name, args);
    }

    if(_tap_todo) {
        fputs(" # TODO", stdout);
        if(*_tap_todo) {
            fputc(' ', stdout);
            fputs(_tap_todo, stdout);
        }
    }

    fputc('\n', stdout);
}

#define _TAP_OK(success, name) do { \
        va_list args; \
        va_start(args, name); \
        _tap_vok(success, name, args); \
        va_end(args); \
    } while(0)

void tap_ok(int success, const char *name, ...)
{
    _TAP_OK(success, name);
}

void tap_is_float(float got, float expected, float delta, const char *name, ...)
{
    float diff = fabs(expected - got);
    int match = diff < delta;
    _TAP_OK(match, name);
    if(!match) {
        tap_diag("expected '%f'", expected);
        tap_diag("got      '%f'", got);
        tap_diag("delta    '%f'", diff);
        tap_diag("allowed  '%f'", delta);
    }
}

void tap_is_int(int got, int expected, const char *name, ...)
{
    int match = got == expected;
    _TAP_OK(match, name);
    if(!match) {
        tap_diag("expected '%d'", expected);
        tap_diag("got      '%d'", got);
    }
}

void tap_is_str(const char *got, const char *expected, const char *name, ...)
{
    int match;
    if(got && expected) {
        match = (strcmp(got, expected) == 0);
    } else {
        match = (got == expected);
    }
    _TAP_OK(match, name);
    if(!match) {
        tap_diag("expected '%s'", expected);
        tap_diag("got      '%s'", got);
    }
}

#endif /* __TAP_C */
