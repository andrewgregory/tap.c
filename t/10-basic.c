#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>

#include "../tap.c"

static char outfile[] = "tmp_10-basic_XXXXXX";
static char expfile[] = "expected/10-basic.t.out";

static void test(void) {
    tap_plan(0);
    tap_plan(1);
    tap_plan(2);

    tap_skip_all(NULL);
    tap_skip_all("foo %s", "bar");

    tap_skip(0, NULL);
    tap_skip(1, "foo %s", "bar");
    tap_skip(2, NULL);

    tap_diag("foo");
    tap_diag("foo %s", "bar");

    tap_bail(NULL);
    tap_bail("foo %s", "bar");

    tap_ok(0, NULL);
    tap_ok(1, NULL);
    tap_ok(0, "%s", "foo");
    tap_ok(1, "%s", "foo");

    tap_is_str("foo", "foo", NULL);
    tap_is_str("foo", "bar", NULL);
    tap_is_str("foo", "foo", "foo %s", "bar");
    tap_is_str("foo", "bar", "foo %s", "bar");
    tap_is_str(NULL, NULL, NULL);
    tap_is_str(NULL, "bar", NULL);

    tap_is_int(1, 1, NULL);
    tap_is_int(1, 0, NULL);
    tap_is_int(1, 1, "foo %s", "bar");
    tap_is_int(1, 0, "foo %s", "bar");

    tap_is_float(1.0, 1.0, 0.1, NULL);
    tap_is_float(1.0, 2.0, 0.1, NULL);
    tap_is_float(1.0, 1.0, 0.1, "foo %s", "bar");
    tap_is_float(1.0, 2.0, 0.1, "foo %s", "bar");

    tap_done_testing();
}

static void run(void) {
    test();
    tap_todo("FOO TODO");
    test();
}

int main(void) {
    int sfd, efd, ofd, ret = 0;
    FILE *efile, *ofile;
    char ebuf[50], obuf[50];
    int line = 0;

    /* redirect output to temp file */
    ofd = mkstemp(outfile);
    sfd = dup(STDOUT_FILENO);
    efd = dup(STDERR_FILENO);
    dup2(ofd, STDOUT_FILENO);
    dup2(ofd, STDERR_FILENO);
    close(ofd);

    run();

    /* flush and restore output */
    fflush(stdout);
    fflush(stderr);
    dup2(sfd, STDOUT_FILENO);
    dup2(efd, STDERR_FILENO);
    close(sfd);
    close(efd);

    efile = fopen(expfile, "r");
    ofile = fopen(outfile, "r");

    puts("1..1");
    while(1) {
        char *eres = fgets(ebuf, 50, efile);
        char *ores = fgets(obuf, 50, ofile);
        ++line;
        if(ores == NULL && eres == NULL) {
            puts("ok 1 - output matches expected");
            break;
        } else if(ores && eres && strcmp(ebuf, obuf) == 0) {
            /* match */
        } else {
            fprintf(stdout, "not ok 1 - mismatch beginning on line %d\n", line);
            fprintf(stderr, "    #      got: '%s'\n", obuf);
            fprintf(stderr, "    # expected: '%s'\n", ebuf);
            ret = 1;
            break;
        }
    }

    fclose(efile);
    fclose(ofile);

    if(getenv("TAP_SAVE")) {
        fprintf(stderr, "%s output saved to %s\n", __FILE__, outfile);
    } else {
        unlink(outfile);
    }

    return ret;
}
