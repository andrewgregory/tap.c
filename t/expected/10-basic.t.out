1..0
1..1
1..2
1..0 # SKIP
1..0 # SKIP foo bar
ok 1 # SKIP foo bar
ok 2 # SKIP
ok 3 # SKIP
# foo
# foo bar
Bail out!
Bail out! foo bar
not ok 4
#   Failed test at 10-basic.c line 29.
ok 5
not ok 6 - foo
#   Failed test at 10-basic.c line 31.
ok 7 - foo
ok 8
not ok 9
#   Failed test at 10-basic.c line 35.
#          got: 'foo'
#     expected: 'bar'
ok 10 - foo bar
not ok 11 - foo bar
#   Failed test at 10-basic.c line 37.
#          got: 'foo'
#     expected: 'bar'
ok 12
not ok 13
#   Failed test at 10-basic.c line 39.
#          got: '(null)'
#     expected: 'bar'
ok 14
not ok 15
#   Failed test at 10-basic.c line 42.
#          got: '1'
#     expected: '0'
ok 16 - foo bar
not ok 17 - foo bar
#   Failed test at 10-basic.c line 44.
#          got: '1'
#     expected: '0'
ok 18
not ok 19
#   Failed test at 10-basic.c line 47.
#          got: '1.000000'
#     expected: '2.000000'
#        delta: '1.000000'
#      allowed: '0.100000'
ok 20 - foo bar
not ok 21 - foo bar
#   Failed test at 10-basic.c line 49.
#          got: '1.000000'
#     expected: '2.000000'
#        delta: '1.000000'
#      allowed: '0.100000'
1..21
1..0
1..1
1..2
1..0 # SKIP
1..0 # SKIP foo bar
ok 22 # SKIP foo bar
ok 23 # SKIP
ok 24 # SKIP
# foo
# foo bar
Bail out!
Bail out! foo bar
not ok 25 # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 29.
ok 26 # TODO FOO TODO
not ok 27 - foo # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 31.
ok 28 - foo # TODO FOO TODO
ok 29 # TODO FOO TODO
not ok 30 # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 35.
#          got: 'foo'
#     expected: 'bar'
ok 31 - foo bar # TODO FOO TODO
not ok 32 - foo bar # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 37.
#          got: 'foo'
#     expected: 'bar'
ok 33 # TODO FOO TODO
not ok 34 # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 39.
#          got: '(null)'
#     expected: 'bar'
ok 35 # TODO FOO TODO
not ok 36 # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 42.
#          got: '1'
#     expected: '0'
ok 37 - foo bar # TODO FOO TODO
not ok 38 - foo bar # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 44.
#          got: '1'
#     expected: '0'
ok 39 # TODO FOO TODO
not ok 40 # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 47.
#          got: '1.000000'
#     expected: '2.000000'
#        delta: '1.000000'
#      allowed: '0.100000'
ok 41 - foo bar # TODO FOO TODO
not ok 42 - foo bar # TODO FOO TODO
#   Failed (TODO) test at 10-basic.c line 49.
#          got: '1.000000'
#     expected: '2.000000'
#        delta: '1.000000'
#      allowed: '0.100000'
1..42
